### PR TITLE
refactor: Moved code of systems and components from Main to separate …

### PR DESCRIPTION
**Pull request by vladjov00**
…files + modified gravity system

Main.cpp now includes GameSpecificComponents.h and GameSpecificSystems.h
- GameSpecificComponents.h: Contains definitions of constants and components with no behaviour
- GameSpecificSystems.h: Includes .h files of all non-Dagger Systems used for SWMG

Gravity now takes into account player's vertical velocity and updates it every tick, so which means that gravitational pull should feel more natural now.

### CODE REVIEW

File: include/GameSpecificComponents.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#include "Prelude.h"
+
+#define SCREEN_WIDTH 800
+#define SCREEN_HEIGHT 600
+#define SPEED_MOD 300.0f
+#define GRAVITY_MOD 400.0f
+#define JUMP_MOD 600.0f
+
+struct Player {
+	bool is_grounded;
```
Comment (gavrilovmiroslav) [inline include/GameSpecificComponents.h:12]: Please use Definitions.h to be more specific about your types. This is useful as both an exercise for when you get a job in a big company and it has its own type-definitions, and as a real-world exercise in precision under stress. In these cases, rewrite this as `Bool`, `F32` and `F64` (for double). Below, there's `const char*` which is `LiteralString`. 

Pointers serve several functions in C++, so we want to be really specific about our intent: we can have a `RawPtr<T>` if we want a real pointer to something. We can use `SharedPtr` for shared ownership (most cases), `UniquePtr` for unique ownership, and `StaticArray<T, N>` to define a static array of type T and size N, a `DynamicArray<T>` to define a vector, and a `SmallBuffer<T>`  for when you know you have like... 10 elements; also `Ref<T>` for `T&`, and a lot of other things. All of this is in Definitions.h, use it often and get used to it!

**Review by gavrilovmiroslav — CHANGES_REQUESTED** [review]

File: include/GameSpecificComponents.h (hunk 1 above)
Comment (gavrilovmiroslav) [reply include/GameSpecificComponents.h:12]: Also, our code is written in the ANSI C++ style, and not the Kernighan & Ritchie style, meaning that `{}` go on their own separate lines.

File: include/Pickup.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+#include "Definitions.h"
+#include "Access.h"
+#include "GameSpecificComponents.h"
+
+struct PickupSystem
+	: public ecs::System
+	, public MutAccessGroupStorage<Player, Position>
+	, public MutAccessGroupStorage<Pickup, Position>
+{
+	using QueryPlayers 	= MutAccessGroupStorage<Player, Position>;
```
Comment (gavrilovmiroslav) [inline include/Pickup.h:12]: weird spacing, no need to align or anything like that, let your IDE do the hard work :)

File: src/Combat.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#include "Combat.h"
+
+void CombatSystem::on_tick()
+{
+    for(auto&& [attack_entity, attack, attack_pos, sprite, flip] : QueryAttacks :: access_storage().each()){
```
Comment (gavrilovmiroslav) [inline src/Combat.cpp:5]: All the `{` go in their own line, make it more readable where the blocks start and end

Comment (gavrilovmiroslav) [inline src/Combat.cpp:5]: `QueryAttacks::access_storage()` goes without spaces, as does any `a.b`, `a::b` etc.

File: src/Combat.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        }
+        else{
+            attack_pos.xy.x -= attack.speed * Time:: delta_time();
+            if(attack_pos.xy.x < attack.range){
+                despawn(attack_entity);
+            }
+        }
+
+        for(auto&& [player_entity, player, player_pos] : QueryPlayers :: access_storage().each()){
+            if (	abs(player_pos.xy.x - attack_pos.xy.x) <= attack.radius
```
Comment (gavrilovmiroslav) [inline src/Combat.cpp:20]: if you want to align these, i'd rather go for aligning with the start of the line but eh, it might be okay

Comment (gavrilovmiroslav) [reply src/Combat.cpp:20]: whatever you choose, choose consistently!

File: src/Platform.cpp [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#include "Platform.h"
+
+void PlatformSystem::on_tick()
+{
+    for (auto&& [player_entity, player, player_pos] : QueryPlayers :: access_storage().each())
+    {
+        bool on_platform = false;
+        for (auto&& [platform_entity, platform, platform_pos] : QueryPlatforms :: access_storage().each())
+        {
+            if(	player_pos.xy.x <= platform_pos.xy.x + platform.width / 2 + 11
```
Comment (gavrilovmiroslav) [inline src/Platform.cpp:10]: I like this style a lot more, so please choose one and adhere to it!

File: include/GameSpecificComponents.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "Prelude.h"
+
+#define SCREEN_WIDTH 800
```
[the remark this answers is not in the document]
Comment (vladjov00) [reply include/GameSpecificComponents.h:5]: I have no idea how the dagger.ini file works, how do other files see and use values defined there?

**Review by jormundur00 — COMMENTED** [review]
Good and very needed refactoring. Gravity files could maybe be renamed into PhysicSystems files, in case of a decision to expand on that system with more physics, but we will cross that bridge when we get there.

**Review by jormundur00 — APPROVED** [review]

**Pull request merged**